### PR TITLE
Semicoarsening nodal

### PR DESCRIPTION
**Pull request by b94209002**
Semicoarsening for nodal solver.

### CODE REVIEW

File: Src/LinearSolvers/MLMG/AMReX_MLNodeLaplacian.cpp [hunk 1]
```diff
@@ -1089,6 +1089,17 @@ MLNodeLaplacian::restriction (int amrlev, int cmglev, MultiFab& crse, MultiFab&
 
     const auto& stencil = m_stencil[amrlev][cmglev-1];
 
+    bool regular_coarsening = true; int idir = 2;
+    if (amrlev == 0 and cmglev > 0) {
+        regular_coarsening = mg_coarsen_ratio_vec[cmglev-1] == mg_coarsen_ratio;
+        IntVect ratio = (amrlev > 0) ? IntVect(2) : mg_coarsen_ratio_vec[cmglev-1];
```
Comment (WeiqunZhang) [inline Src/LinearSolvers/MLMG/AMReX_MLNodeLaplacian.cpp]: This is inside `if (amrlev == 0 and ..)` so it's not possible `amrlev > 0`.

Comment (b94209002) [reply Src/LinearSolvers/MLMG/AMReX_MLNodeLaplacian.cpp]: Sure, it can be left with just cmglev > 0.

File: Src/LinearSolvers/MLMG/AMReX_MLNodeLaplacian.cpp [hunk 2]
```diff
@@ -1140,6 +1161,18 @@ MLNodeLaplacian::interpolation (int amrlev, int fmglev, MultiFab& fine, const Mu
 
     const iMultiFab& dmsk = *m_dirichlet_mask[amrlev][fmglev];
 
+    bool regular_coarsening = true; int idir = 2;
+    if (amrlev == 0 and fmglev > 0) {
+        regular_coarsening = mg_coarsen_ratio_vec[fmglev] == mg_coarsen_ratio;
+	IntVect ratio = (amrlev > 0) ? IntVect(2) : mg_coarsen_ratio_vec[fmglev];
```
Comment (WeiqunZhang) [inline Src/LinearSolvers/MLMG/AMReX_MLNodeLaplacian.cpp]: The test on amrlev is unnecessary.  Also the indentation seems to be off.  If this is because of tabs, please use four white spaces instead of tab.

File: Src/LinearSolvers/MLMG/AMReX_MLNodeLaplacian.cpp [hunk 3]
```diff
@@ -1357,6 +1400,11 @@ MLNodeLaplacian::Fsmooth (int amrlev, int mglev, MultiFab& sol, const MultiFab&
 
     const iMultiFab& dmsk = *m_dirichlet_mask[amrlev][mglev];
 
+    bool regular_coarsening = true;
+    if (amrlev == 0 and mglev > 0) {
+        regular_coarsening = mg_coarsen_ratio_vec[mglev-1] == mg_coarsen_ratio;
+    }
+
```
Comment (WeiqunZhang) [inline Src/LinearSolvers/MLMG/AMReX_MLNodeLaplacian.cpp]: This is used in a lot of places now.  Maybe we can put it in function say `bool is_regular_coarsening(amrlev, fine_mglev)`.

File: Tutorials/LinearSolvers/NodalPoisson/GNUmakefile [hunk 4]
```diff
@@ -5,7 +5,7 @@ USE_OMP  = FALSE
 
 COMP = gnu
 
-DIM = 3
+DIM = 2
```
Comment (WeiqunZhang) [inline Tutorials/LinearSolvers/NodalPoisson/GNUmakefile]: Please change this back.

File: Tutorials/LinearSolvers/NodalPoisson/MyTest.H [hunk 5]
```diff
@@ -37,6 +37,13 @@ private:
 
     int gpu_regtest = 0;
 
+
+    bool agglomeration = true;
+    bool consolidation = true;
+    bool semicoarsening = false;
+    int max_semicoarsening_level = 0;
+
+
```
Comment (WeiqunZhang) [inline Tutorials/LinearSolvers/NodalPoisson/MyTest.H]: Are these being used?

File: Src/LinearSolvers/MLMG/AMReX_MLNodeLap_2D_K.H [hunk 6]
```diff
@@ -502,6 +502,146 @@ void mlndlap_gauss_seidel_aa (Box const& bx, Array4<Real> const& sol,
...
+        idir = 0;
+        ilen = hi.x - lo.x + 1;
+    }
+
+    if (ilen > 32) amrex::Abort("mlndlap_gauss_seidel_with_line_solve_aa is hard-wired to be no longer than 32");
+
+    Array1D<Real,0,31> a_ls,b_ls,c_ls,u_ls,r_ls,gam;
+
+    if (idir == 1) {
+	AMREX_PRAGMA_SIMD
```
Comment (WeiqunZhang) [inline Src/LinearSolvers/MLMG/AMReX_MLNodeLap_2D_K.H]: This should be removed.

File: Src/LinearSolvers/MLMG/AMReX_MLNodeLap_2D_K.H [hunk 7]
```diff
@@ -502,6 +502,146 @@ void mlndlap_gauss_seidel_aa (Box const& bx, Array4<Real> const& sol,
...
+            }
+            tridiagonal_solve(a_ls, b_ls, c_ls, r_ls, u_ls, gam, ilen);
+
+            for (int j = lo.y; j <= hi.y; ++j)
+            {
+            	sol(i,j,k) = u_ls(j-lo.y);
+            }
+        }
+    } else if (idir == 0) {
+        AMREX_PRAGMA_SIMD
```
Comment (WeiqunZhang) [inline Src/LinearSolvers/MLMG/AMReX_MLNodeLap_2D_K.H]: This should be removed.

File: Src/LinearSolvers/MLMG/AMReX_MLNodeLap_3D_K.H [hunk 8]
```diff
@@ -1135,6 +1135,263 @@ void mlndlap_gauss_seidel_aa (Box const& bx, Array4<Real> const& sol,
...
+    if (ilen > 32) amrex::Abort("mlndlap_gauss_seidel_with_line_solve_aa is hard-wired to be no longer than 32");
+
+    Array1D<Real,0,31> a_ls,b_ls,c_ls,u_ls,r_ls,gam;
+
+
+    if ( idir == 2 ) 
+    {
+        for (int j = lo.y; j <= hi.y; ++j) 
+	{
+            AMREX_PRAGMA_SIMD
```
Comment (WeiqunZhang) [inline Src/LinearSolvers/MLMG/AMReX_MLNodeLap_3D_K.H]: This should be removed.

File: Src/LinearSolvers/MLMG/AMReX_MLNodeLap_3D_K.H [hunk 9]
```diff
@@ -1135,6 +1135,263 @@ void mlndlap_gauss_seidel_aa (Box const& bx, Array4<Real> const& sol,
...
+	      	    sol(i,j,k) = u_ls(k-lo.z);
+		}
+	    }
+        }
+    }
+    else if (idir == 1)
+    {
+        for (int i = lo.x; i <= hi.x; ++i)
+        {
+            AMREX_PRAGMA_SIMD
```
Comment (WeiqunZhang) [inline Src/LinearSolvers/MLMG/AMReX_MLNodeLap_3D_K.H]: This should be removed.  Also are there any reasons the i-loop is outside the k-loop?  Generally speaking we want k-loop to be outside i-loop because the arrays are in Fortran order.

File: Src/LinearSolvers/MLMG/AMReX_MLNodeLap_3D_K.H [hunk 10]
```diff
@@ -1135,6 +1135,263 @@ void mlndlap_gauss_seidel_aa (Box const& bx, Array4<Real> const& sol,
...
+                    sol(i,j,k) = u_ls(j-lo.y);
+                }
+	    }
+	}
+    } 
+    else if (idir == 0) 
+    {
+        for (int j = lo.y; j <= hi.y; ++j)
+        {
+            AMREX_PRAGMA_SIMD
```
Comment (WeiqunZhang) [inline Src/LinearSolvers/MLMG/AMReX_MLNodeLap_3D_K.H]: This should be removed.

File: Src/LinearSolvers/MLMG/AMReX_MLNodeLap_3D_K.H (hunk 9 above)
Comment (WeiqunZhang) [reply Src/LinearSolvers/MLMG/AMReX_MLNodeLap_3D_K.H]: If there are no reasons for the loop order chosen, please switch the orders in other places too.

File: Src/LinearSolvers/MLMG/AMReX_MLNodeLap_3D_K.H [hunk 11]
```diff
@@ -1135,6 +1135,263 @@ void mlndlap_gauss_seidel_aa (Box const& bx, Array4<Real> const& sol,
...
+    Array1D<Real,0,31> a_ls,b_ls,c_ls,u_ls,r_ls,gam;
+
+
+    if ( idir == 2 ) 
+    {
+        for (int j = lo.y; j <= hi.y; ++j) 
+	{
+            AMREX_PRAGMA_SIMD
+            for (int i = lo.x; i <= hi.x; ++i) 
+	    {
```
Comment (WeiqunZhang) [inline Src/LinearSolvers/MLMG/AMReX_MLNodeLap_3D_K.H]: If you look this on github, the indention here and many other places are not right.  Please fix them (by avoid tabs).  It's better you fix it than someone else fixes it later, because it makes the history hard to follow.

**Comment by b94209002** [comment]
Try to fix everything that you mentioned. If that is not quite clean, I can clean the commits out and make a new one.

**Review by WeiqunZhang — APPROVED** [review]
